networkx/algorithms/bipartite/matching.py

```diff
@@ -582,3 +585,310 @@ def minimum_weight_full_matching(G, top_nodes=None, weight="weight"):
     # add the ones from right to left as well.
     d.update({v: u for u, v in d.items()})
     return d
+
+
+def _neighbours_of_set(G, node_set):
+    r"""Returns a set of the neighbours of a given set of nodes `node_set` in the bipartite graph `G`.
+    Denoted by :math:`N_Gnode_set`.
+
+    `G` is an undirected bipartite NetworkX graph.
+
+    `node_set` is a set of nodes in `G`.
+
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> _neighbours_of_set(G, {})
+    set()
+    >>> _neighbours_of_set(G, {1, 2})
+    {3, 4, 5}
+
+    >>> G = nx.Graph([(0, 4), (1, 5), (2, 6)])
+    >>> _neighbours_of_set(G, {0, 1})
+    {4, 5}
+
+    >>> G=nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+    >>> _neighbours_of_set(G, {0, 1})
+    {3, 4}
+
+    >>> _neighbours_of_set(G, {4})
+    {0, 1, 2}
+    """
+    ret_set = {}
+    for node in node_set:
+        ret_set.update(G[node])
+
+    return set(ret_set)
+
+
+def _M_alternating_sequence(G, M, *, top_nodes=None):
+    r"""Generates a M-alternating-sequence of a bipartite graph `G` with regard to a matching M.
+
+    Two sets are calculated with the following recursive definition:
+
+    X_0 = nodes in G unmatched by M.
+    for i>=1:
+        Latex:
+            Y_i = N_{G\M}(X_{i-1})
+            X_i = N_M(Y_i)
+        English:
+            Y_i -> neighbours of X_{i-1} that were not matched by M group-subtracting all the previous Y_k's we've seen
+            X_i -> neighbours of Y_i that were matched by M
+
+    The iterative pattern stops when either Y_k or X_k are an empty set for a certain k, forever staying an empty set from k onward
+    the alternating sequece is proven to always stop in the article.
+
+    >>> G = nx.complete_bipartite_graph(3,3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> _M_alternating_sequence(G, M)
+    ((), ())
+
+    >>> G = nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3, 3:0, 1:4, 4:1}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}), ({4},))
+
+    >>> G = nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+    >>> M = {0:3,3:0,4:1,1:4}
+    >>> _M_alternating_sequence(G, M)
+    (({2}, {1}, {0}), ({4}, {3}))
+
+    >>> G = nx.Graph([(0,6),(6,0),(1,6),(6,1),(1,7),(7,1),(2,6),(6,2),(2,8),(8,2),(3,9),(9,3),(3,6),(6,3),(4,8),(8,4),(4,7),(7,4),(5,9),(9,5)])
+    >>> M = {0:6,6:0,1:7,7:1,2:8,8:2,3:9,9:3}
+    >>> _M_alternating_sequence(G, M)
+    (({4, 5}, {1, 2, 3}, {0}), ({8, 9, 7}, {6}))
+
+    """
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_0 = set(X) - set(M.keys())
+    if X_0 == set():
+        return (), ()
```

Comment:
I have question about the algorithms (sorry I have not read the paper yet!)
If I switch one node from one set to another bipartite set then the this would not that work right?
``` python
In [97]: G = nx.Graph([(0,3), (0,4), (1,4), (2,0)])

In [98]: M = nx.bipartite.maximum_matching(G)

In [99]: _M_alternating_sequence(G, M)
Out[99]: ((), ())

In [100]: G = nx.Graph([(0,3), (0,4), (1,4), (2,4)])

In [101]: M = nx.bipartite.maximum_matching(G)

In [102]: _M_alternating_sequence(G, M)
Out[102]: (({2}, {1}), ({4},))
``` 
should the output of 99 and 102 be the same?

I moved node `2` to the other side, should I be getting something like `(({1}, ), ({4}, {2}))` ?

Reply:
When you say "would not work" I'm not sure what exactly you mean.
when you took 2 and moved it to the other bipartite set and connected it to 4 you just created a different graph, the resulting M-alternating-sequence may be different.
The output in this algorithm is a partition of the two bipartite sets in the graph into disjoint sets.
This is similar to an M-alternating path.
All the algorithms we implemented are implementations of the ones written in pseudo-code or in words in the article.
When you get the output '((), ())' this happens because the M-alternating-sequence is a sequence starting from vertices unmatched by M, so if M matches all the vertices in X (where X,Y are the bipartite sets of G) the algorithm terminates immediately and the M-alternating-sequence is empty.
'I moved node 2 to the other side, should I be getting something like (({1}, ), ({4}, {2})) ?' this depends, the output of the algorithm depends on the matching returned (keep in mind there may be multiple different maximum matchings for a given bipartite graph) as well as the edges in the graph.
In general I can say that it may not be of utmost importance for you to understand the calculation of M-alternating-sequences since this is only a helper function used for calculating the envy-free bipartite matching which is the subject of this PR, and the contribution in this PR is mainly calculating these envy-free bipartite matching, I didn't mean for these helper functions to be made public in the library (but I'm not opposed to it).

Anyway, I'd be glad to offer any help in further clarifying how these algorithms work if you have any questions (as well as further explaining about what you asked above).
If you have time/want to you can read the article, but it's really not mandatory in order to obtain a higher-level understanding of these algorithms.

I'm not sure if this is a conventional custom, but if you want we can maybe schedule a virtual meeting so we can chat and I can explain the algorithms thoroughly.